source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -374,47 +372,50 @@ ir::Instruction* FoldSpecConstantOpAndCompositePass::DoVectorShuffle(
   assert(result_vec_type &&
          "The result of VectorShuffle must be of type vector");
 
+  // A temporary buffer of new null constants to be build in case any of the
+  // vector operands are declared with OpConstantNull.
+  std::vector<std::unique_ptr<analysis::Constant> > null_constants_to_build;
+
   // Get a concatenated vector of scalar constants. The vector should be built
   // with the components from the first and the second operand of VectorShuffle.
-  std::vector<const analysis::Const*> concatenated_components;
+  std::vector<const analysis::Constant*> concatenated_components;
   for (uint32_t i : {1, 2}) {
     assert(inst->GetInOperand(i).type == SPV_OPERAND_TYPE_ID &&
            "The vector operand must have a SPV_OPERAND_TYPE_ID type");
     uint32_t operand_id = inst->GetSingleWordInOperand(i);
-    analysis::Const* operand_const = FindRecordedConst(operand_id);
+    analysis::Constant* operand_const = FindRecordedConst(operand_id);
     if (!operand_const) return nullptr;
     const analysis::Type* operand_type = operand_const->type();
     assert(operand_type->AsVector() &&
            "The first two operand of VectorShuffle must be of vector type");
-    if (analysis::VectorConst* vec_const = operand_const->AsVectorConst()) {
+    if (analysis::VectorConstant* vec_const =
+            operand_const->AsVectorConstant()) {
       // case 1: current operand is a non-null vector constant.
       for (uint32_t j = 0; j < vec_const->NumComponents(); j++) {
         concatenated_components.push_back(vec_const->GetComponent(j));
       }
-    } else if (operand_const->AsNullConst()) {
+    } else if (operand_const->AsNullConstant()) {
       // case 2: current operand is a null vector constant. Create a temporary
       // null scalar constant as the component.
       const analysis::Type* component_type =
           operand_type->AsVector()->element_type();
-      auto component_const = CreateConst(component_type, {});
+      null_constants_to_build.emplace_back(CreateConst(component_type, {}));
       // Append the null scalar consts to the concatenated components
       // vector.
       concatenated_components.insert(concatenated_components.end(),
                                      operand_type->AsVector()->element_count(),
-                                     component_const.get());
-      // Insert an instruction for the component constant.
-      // TODO(qining): Here we may insert an instruction while return a null
-      // pointer. In that case, the process() will return false and assumes the
-      // module is not modified, but actually it is modified. Need to fix it.
-      BuildInstructionAndAddToModule(std::move(component_const), pos);
+                                     null_constants_to_build.back().get());
     } else {
       // no other valid cases
       return nullptr;
     }
   }
+  // Create null component constants if there are any.
+  for (auto& c : null_constants_to_build) {
```

Comment:
You can reuse the same null constant multiple times.  So you only need a pointer to single constant, it's nullptr if you never need to make one.